File: lib/py/uapi/internal/generation/ConstructRandomStruct.py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from uapi.RandomGenerator import RandomGenerator
    from uapi.internal.types.UFieldDeclaration import UFieldDeclaration
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration


def construct_random_struct(reference_struct: dict[str, 'UFieldDeclaration'],
                            starting_struct: dict[str, object],
                            include_optional_fields: bool,
                            randomize_optional_fields: bool,
                            type_parameters: list['UTypeDeclaration'],
                            random_generator: 'RandomGenerator') -> dict[str, object]:

    sorted_reference_struct = sorted(
        reference_struct.items(), key=lambda x: x[0])

    obj = {}
    for field_name, field_declaration in sorted_reference_struct:
        blueprint_value = starting_struct.get(field_name)
        use_blueprint_value = field_name in starting_struct
        type_declaration = field_declaration.type_declaration

        if use_blueprint_value:
            value = type_declaration.generate_random_value(blueprint_value, use_blueprint_value,
                                                           include_optional_fields, randomize_optional_fields,
                                                           type_parameters, random_generator)
        else:
            if not field_declaration.optional:
                value = type_declaration.generate_random_value(None, False,
                                                               include_optional_fields, randomize_optional_fields,
                                                               type_parameters, random_generator)
            else:
                if not include_optional_fields or (randomize_optional_fields and random_generator.next_boolean()):
                    continue
                value = type_declaration.generate_random_value(None, False,
                                                               include_optional_fields, randomize_optional_fields,
                                                               type_parameters, random_generator)

        obj[field_name] = value

    return obj

File: lib/java/src/main/java/uapi/internal/generation/ConstructRandomStruct.java
package uapi.internal.generation;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Map;
import java.util.TreeMap;

import uapi.RandomGenerator;
import uapi.internal.types.UFieldDeclaration;
import uapi.internal.types.UTypeDeclaration;

public class ConstructRandomStruct {
    static Map<String, Object> constructRandomStruct(
            Map<String, UFieldDeclaration> referenceStruct, Map<String, Object> startingStruct,
            boolean includeOptionalFields, boolean randomizeOptionalFields,
            RandomGenerator randomGenerator) {

        final var sortedReferenceStruct = new ArrayList<>(referenceStruct.entrySet());
        Collections.sort(sortedReferenceStruct, (e1, e2) -> e1.getKey().compareTo(e2.getKey()));

        final var obj = new TreeMap<String, Object>();
        for (final var field : sortedReferenceStruct) {
            final var fieldName = field.getKey();
            final var fieldDeclaration = field.getValue();
            final var blueprintValue = startingStruct.get(fieldName);
            final var useBlueprintValue = startingStruct.containsKey(fieldName);
            final UTypeDeclaration typeDeclaration = fieldDeclaration.typeDeclaration;

            final Object value;
            if (useBlueprintValue) {
                value = typeDeclaration.generateRandomValue(blueprintValue, useBlueprintValue,
                        includeOptionalFields, randomizeOptionalFields, randomGenerator);
            } else {
                if (!fieldDeclaration.optional) {
                    value = typeDeclaration.generateRandomValue(null, false,
                            includeOptionalFields, randomizeOptionalFields, randomGenerator);
                } else {
                    if (!includeOptionalFields || (randomizeOptionalFields && randomGenerator.nextBoolean())) {
                        continue;
                    }
                    value = typeDeclaration.generateRandomValue(null, false,
                            includeOptionalFields, randomizeOptionalFields, randomGenerator);
                }
            }

            obj.put(fieldName, value);
        }
        return obj;
    }
}
